### PR TITLE
Packaging up so can be installed via yarn or npm.

### PR DESCRIPTION
**Pull request by martin-bucinskas**
# Changes
Created this into a package form so you can install it through `yarn` or `npm`.

 - Created a package.json which contains a postinstall script to copy over the dot files and jsig-allow.json file.
 - jsig-allow.json file now contains an entry for itself to prevent errors when adding additional packages.

# Why
This reduces the amount of items required in the root of the project when installing, with this being mostly self contained in the `node_modules` directory.
Also makes it easier to install, without having the need to dowload an archive, unpackage it and copy over to the project.

### CODE REVIEW

**Comment by ashward** [comment]
Thanks Martin. That's a really nice idea, but the trouble is that this script needs to be there before any packages are installed. A couple of scenarios I'm thinking:
* Someone has checked out an existing project with jsig installed. There won't be a node_modules folder, so jsig won't be available until `npm/yarn install` is invoked.
* Someone deletes the node_modules folder. Much like above the jsig script won't be available.
* Someone checks out a project without jsig installed in it. The first thing they want to do is install jsig, so they do a `yarn add`. Unfortunately this also seems to do an `install` of all the other packages, which will be done without jsig in place.

My original thinking was to have a shell script to install jsig. However, I'm now thinking we could:
* Have it installable as a shell script so that it can be installed without yarn (which installs the scripts outside the node_modules folder, in the .jsig directory) - this is so that it can be added to existing projects without doing a yarn/npm install. This could also potentially add it to the package.json.
* Also have it available as a package (which can be used for new projects) which also installs the script outside the node_modules folder (same as the bash script)
* Upgrades could then be done using the existing package manager.

So actually I think that scenario is very similar to what you've got here, except the scripts are copied (and used) outside the node_modules folder.

Sorry about the ramble! I hope that makes sense!

**Comment by ashward** [comment]
So maybe if the post-install script also copied the javascript files to the `/.jsig` directory and the `.yarnrc` and `.npmrc` files pointed there (rather than into the node_modules dir) then that would work...

File: jsig-allow.json [hunk 1]
```diff
@@ -0,0 +1,5 @@
+{
+    "allow": [
+        {"jsig":"0.1.5","path":"node_modules/jsinstallguard","v":"0.1.5","cmd":"cp ./{.npmrc,.yarnrc,jsig-allow.json} ../../"}
```
Comment (ashward) [inline jsig-allow.json]: I wonder if we add this trust within the code itself (maybe just add it to the allow array in the code). That way if someone has an existing jsig-allow files then we can just leave it alone.

File: package.json [hunk 2]
```diff
@@ -0,0 +1,22 @@
+{
+  "name": "jsinstallguard",
+  "version": "0.1.5",
+  "description": "Small security wrapper around the package manager which intercepts any preistall and postinstall scripts.",
+  "main": "index.js",
+  "scripts": {
+    "postinstall": "cp ./{.npmrc,.yarnrc,jsig-allow.json} ../../"
```
Comment (ashward) [inline package.json]: Probably need to consider what happens if someone already has a `.npmrc`, `.yarnrc`, or `jsig-allow.json` file.

**Comment by martin-bucinskas** [comment]
Makes sense, this was just a POC for me :) It's just a pain at the moment to copy stuff over, feels like there should be a much simpler way!

File: jsig-allow.json (hunk 1 above)
Comment (martin-bucinskas) [reply jsig-allow.json]: Yeah, I agree, this was knocked up quickly for POC.

File: package.json (hunk 2 above)
Comment (martin-bucinskas) [reply package.json]: Yup, was thinking the same, gits merge capability would be nice

**Comment by ashward** [comment]
I think it's actually a great idea - better than my original one of just using shell scripts. I think with a couple of tweaks we can get it working!

**Comment by ashward** [comment]
@martin-bucinskas I had a go at putting some extra logic in the postinstall. It's still a bit rought around the edges though.

What do you think?

**Review by martin-bucinskas — COMMENTED** [review]
Looks good to me :) Seems a more refined way which handles dot files properly.

**Comment by ashward** [comment]
I've pushed up some more changes. I think it's working pretty reasonably now.

It'll need an uninstall script too ideally, but I'll open a separate ticket for that.

**Review by ashward — APPROVED** [review]

**Pull request merged**